### ISSUE TITLE
.bulk_then_execute expects its user function to receive a future parameter

### ISSUE THREAD

**Issue by jaredhoberock** [executors/executors#383]
The current proposal describes `.bulk_then_execute` this way:

> If `R` is non-`void`, invokes `rf()` at most once on an unspecified execution agent to produce the value `r`. Invokes `sf()` at most once on an unspecified execution agent to produce the value `s`. When `fut` is ready, creates a group of execution agents of shape `n` which invokes `DECAY_COPY( std::forward<F>(f))(i, fut, r, s)` if `R` is non-`void`, and otherwise invokes `DECAY_COPY( std::forward<F>(f))(i, fut, s)`, at most once for each value of `i` in the range `[0,n)`, with the call to `DECAY_COPY` being evaluated in the thread that called `bulk_then_execute`.

However, in R3, `.bulk_then_execute`'s operational semantics include:

> If `R` is non-`void`, invokes `rf()` on an unspecified execution agent to produce the value `r`. Invokes `sf()` on an unspecified execution agent to produce the value `s`. When `pred` is ready, creates a group of execution agents of shape `n` which invokes `DECAY_COPY( std::forward<F>(f))(i, pred, r, s)` if `R` is non-`void`, and otherwise invokes `DECAY_COPY( std::forward<F>(f))(i, pred, s)`, at most once for each value of `i` in the range `[0,n)`, with the call to `DECAY_COPY` being evaluated in the thread that called `bulk_twoway_execute` [sic].

In this text, `pred` is the value of the predecessor future.

The current text requires `f` to receive the future, rather than the future's ready value, as a parameter. The problem is that it this implies that each execution agent in the group must call `fut.get()` to unwrap its value. If multiple agents do this in parallel, they will race.

We could revert to the older semantics, but this would create an inconsistency with `.then_execute`'s expectations.

**Comment by ccmysen** [comment]
From the email chain, my suggestion would be that we introduce effectively a reducer phase which handles the conversion on this same topic. The promise based proposal allows us to basically break then execute into 3 stages where exception and value reduction are part of the same operation...

my thought is similar to @LeeHowes suggestion, but perhaps to formalize that we are not just processing a single value & exception at a time but to actually allow the intermediate stage between completion of the previous stage and start of the next stage to behave like a full map+reduce where you feed all values to some operation which allows the exceptions and values to be redirected.

this would allow the exceptions to be eaten or passed on with values to certain indices of the bulk operation...effectively the idea is to default to passing exceptions on just like the value (at the same indices), but allow an intermediate stage to eat the exceptions and pass on only values, or even to post-process and create a new shared state to be passed onto the bulk operation.

one possible type of interface for this would be:
```
future<std::pair<int, std::exception_ptr>> f = ex.bulk_then_execute(
   [](
       int n,
       ready_future<int>* input_value,
       ready_future<int>* result,
       int&){
     std::exception_ptr ex = nullptr;
     for (size_t i = 0; i < n; ++i) {
       if(!input_value[i].has_value()) {
           ex = input_value[i].exception();
           break;
        }
     }
     if(ex != nullptr)
     for (size_t i = 0; i < n; ++i) {
        // copy value + exception to everyone
        if (input_value[i].has_value()) {
          // Special dual future with exception and value -- assumes this is useful.
          result[i] = make_ready_future(input_value.value(), ex);
        } else {
          result[i] = make_exceptional_future(ex);
        }
     }
    }
   },
   8,
   source_future,
   []() -> ready_future<int>{ return {0, nullptr}; },
   []{ return 0; });
```

**Comment by ccmysen** [comment]
you could of course do something similar by passing all exceptions and values into a non-bulk then_execute and then creating shared state there which calls bulk_execute.

this of course assumes we have a way of calling a chain like this which makes sense:

bulk_execute().then_execute().bulk_then_execute()

**Comment by LeeHowes** [comment]
The factory functions/resulting types should already allow much of that. If we start requiring actual map-reduce tasks to operate, then we might as well just enqueue followup scalar tasks.

My only concern with doing this the obvious way: passing a ready_future by reference into each task and using the result factory to do the reduction is that we have to launch the full set of tasks in the exception case. A task-graph communicated bypass might be a nice addition.

Another simple change that might help here would be a tweak to the result factory. In https://github.com/executors/executors-impl/blob/master/examples/exception/bulk_twoway_reduce.cpp I added a scalar followup task to unpack the reduction state. If the result factory is also defined to return a ready_future, then that ready_future can be an arbitrarily complex type defined by the operation, for example, to return just the first exception:

```
struct ReducingFuture {
  std::vector<value_type> values_;
  std::vector<exception_ptr> exceptions_;

  value_type get() {
    if(exceptions_.size() > 0) {
      std::rethrow_exception(exceptions_[0]);
    } else {
      value_type acc{};
      for(auto& v : values_) {
        acc += v;
      }
      return acc;
    }
  }
 // has_value(), has_exception(), exception() etc
}
```

We reduce in whatever way we want, but we have an opportunity to also slip some code into the get operation, so that our extra state for reduction purposes that carries both a or many values, and exceptions, is then reduced by the standard ready_future -> ready_future conversion operation in chaining.

**Comment by LeeHowes** [comment]
Ok, one more tweak to this based on thoughts in the e-mail thread, give the shape_factory access to the output. Then it can skip the bulk execution entirely just forwarding the exception to the result.

```
future<std::pair<int, std::exception_ptr>> f = ex.bulk_then_execute(
   [](
       int n,
       ready_future<int>& input_value,
       ReducingFuture<int>& result,
       int&){
      // Code for task
      if(input_value.has_exception()) {
        // clearly this would have to be a concurrency safe vector in practice
        result.exceptions_.push_back(input_value.exception());
        return;
      }
      result.values_[n] = input_value.get();
    }
   },
   8,
   source_future,
   [](int input_shape, 
       const ready_future<int>& input_future, 
       ReducingFuture<int><int>& result_future) {
      // Shape factory modifies the shape based on the input data and the originally requested shape
      // it can optionally update the output
      if(input_future.has_exception()) {
        // Store the exception directly in the result
        result_future.exceptions_.push_back(input_future.exception());
        // Do not dispatch bulk tasks
        return 0;
      } else {
        return input_shape;
      }, 
   []() -> ReducingFuture<int>{ return {0, nullptr}; },
   []{ return 0; });
```

**Comment by jaredhoberock** [comment]
Here's a random idea. I'm not sure if I think it's a good one. But what if we allowed the user function to be overloaded, and specify that the executor invoke a different overload of the user function when the incoming predecessor is exceptional?

To make it concrete, here's an example of how it could work with an inline bulk executor:

    struct inline_bulk_executor
    {
      template<class F, class Future, class RF, class SF>
      future<auto> bulk_then_execute(F f, size_t n,
                                     Future& fut,
                                     RF result_factory, SF shared_factory) const
      {
        auto result = result_factory();
        auto shared = shared_factory();

        try
        {
          // normal case
          auto& predecessor = fut.get();
          for(size_t i = 0; i < n; ++i)
          {
            f(i, predecessor, result, shared);
          }
        }
        catch(...)
        {
          // exceptional case
          auto exception = current_exception();
          for(size_t i = 0; i < n; ++i)
          {
            f(i, exception, result, shared);
          }
        }

        return make_ready_future(result);
      }
    };

If `f` can't be invoked with the predecessor future's associated exception type, then the incoming exception is simply forwarded to the result future.

**Comment by LeeHowes** [comment]
While that would allow me to package two callbacks into F and share state between them, it would make using an auto parameter to the callback hard. It also hides the branch functionality from my control as a library author, which is unfortunate. By the time you've done that, why not just make a trivial ready_future type that references the source future in the case above (or whatever state you need to make it async instead of calling .get()) and pass it into a single callback to give control back to the programmer? There's next to no cost to you as the executor author then.

**Comment by LeeHowes** [comment]
Thinking about this further, your idea is basically what I started the day with at the F2F. It does require a concept, it's just that it is a promise concept. I don't think it is simpler than having a future parameter, and it would probably mean more required data moves that could be hidden under a reference-storing future.

We can have two function overloads in there, or set_value and set_exception to make it concrete, where set_exception is optional if you want to allow value-only promises. The callback is then a promise, just like any other promise, defined only by its calling side.

In that world optimally we'd like to support both returning and sink promises - which differ in that future.continue_with(sink_promise) should not return a future, or returns a future that is immediately ready or is ready when that state of the computation completes. I don't know if passing a sink_promise into then_execute makes any sense.

From the point of view of the future API, we'd have continue_with(Promise), then(value callback), onError(exception callback). The latter two would package a promise up and pass it to then_execute, the former would pass it directly if it is not a sink, and call it lazily if it is.

**Comment by LeeHowes** [comment]
Although to put it in perspective, this would be a ready_future version of the above:
struct inline_bulk_executor
{
  template<class T>
  class ready_future {
     T* value;
     exception_ptr exception;
 
     ready_future(T&& val) : value{val);

     T& get() {
        if(value) return value; else std::rethrow_exception(exception);
      }
     // other trivial methods
  };
  template<class F, class Future, class RF, class SF>
  future<auto> bulk_then_execute(F f, size_t n,
                                 Future& fut,
                                 RF result_factory, SF shared_factory) const
  {
    auto result = result_factory();
    auto shared = shared_factory();

    try
    {
      // normal case
      auto predecessor_value = fut.get();
      auto rf = 
      for(size_t i = 0; i < n; ++i)
      {
        f(i, rf, result, shared);
      }
    }
    catch(...)
    {
      // exceptional case
      auto predecessor_exception = current_exception();
      for(size_t i = 0; i < n; ++i)
      {
        f(i, rf, result, shared);
      }
    }

    return make_ready_future(result);
  }
};

**Comment by LeeHowes** [comment]
Ok, I've played with this a little and chatted with colleagues. Your idea isn't crazy, but it does have flaws. For a start, your code excludes a single auto callback, or at least makes that callback messy (messier even than requiring a Maybe type), because we'd need to do a static check on whether the parameter is an exception_ptr or not. Given that, in practice we'd end up having to always provide a struct with an exception_ptr overload to go with the auto one.

So what if instead we compromise on this and actually give this something resembling Promise's interface? So your code changes to:

```
struct inline_bulk_executor
{
  template<class F, class Future, class RF, class SF>
  future<auto> bulk_then_execute(F f, size_t n,
                                 Future& fut,
                                 RF result_factory, SF shared_factory) const
  {
    auto result = result_factory();
    auto shared = shared_factory();

    try
    {
      // normal case
      auto predecessor_value = fut.get();
      for(size_t i = 0; i < n; ++i)
      {
        f.set_value(i, predecessor_value, result, shared);
      }
    }
    catch(...)
    {
      // exceptional case
      auto predecessor_exception = current_exception();
      for(size_t i = 0; i < n; ++i)
      {
        f.set_exception(i, predecessor_exception, result, shared);
      }
    }

    return make_ready_future(result);
  }
};
```

At the executor level this stays very simple. At the Future of algorithm level obviously we obviously have to package into a little struct, but that is trivial and we could even provide helpers to do that. It avoids requiring a Maybe type. It still works with a Maybe type if we want it to because we can call the same function from the two overloads if we provide it with a type constructible from either.

It also means we can make set_exception optional and SFINAE against it so that the implementation can bypass if it is not provided.

We need to take a little care to separate the Promise and Continuation concepts here - only in that it makes no sense for a promise to return values from its functions. That should be easily resolved if we can get past this design decision.

**Comment by dhollman** [comment]
> your code excludes a single auto callback, or at least makes that callback messy

Although I agree in principle with the rest of what Lee is saying, I think we (myself included) keep forgetting just how much less messy `if constexpr` can make some of these things. If you want to handle exceptions and values in the same generic lambda, all you have to do is pass something like this to `bulk_then_execute`:

```c++
[](auto i, auto&& maybe_value, auto&& result, auto&& shared) {
  if constexpr (is_same_v<decay_t<decltype(maybe_value)>, exception_ptr>) {
    // handle exception, type of maybe_value is (reference to) exception_ptr
  } else {
    // handle value, type of maybe_value is (reference to) the value
  }
};
```

That being said, I think it very much complicates the signature and description of `then_execute` and friends. We'd essentially be doing the same thing as Lee is proposing, except using overloads instead of method names, which feels a bit less clear to me.

IMHO it's just syntax either way (especially once `if constexpr` is considered), so we should just do whatever increases consensus the most.

**Comment by jaredhoberock** [comment]
`.bulk_then_execute` needs to be able to receive lambdas and function pointers as the user function argument. If disambiguating the result/exception cases is necessary, I was thinking an option could be to use a tag type e.g. `exception_arg_t` (cf. `allocator_arg_t`). Or just use `constexpr if` like David suggested.

**Comment by LeeHowes** [comment]
Yes it can, but it makes auto useless because as soon as we put auto in we lose that conversion and have to add a static if to disambiguate. That's a bad user experience for the sake of not defining a single parameter concept. In that case, why not just define that the expected parameter is some sort of expected concept and make it clear to the code that that is always the case?

I don't really understand why bulk_then_execute needs to receive lambdas and function pointers directly as the argument. This is a low-level interface, we've already agreed elsewhere that we don't want to simplify the result and shared factories, which are difficult enough to use.

How is wrapping the user function trivially in the future or algorithm code like:
```
ex.bulk_then_execute(make_promise_callable_from_value_callback([](int idx, expected_ref<int> maybe_int, auto& result, auto& shared)
{...}),
n, predecessor, result_factory, shared_factory
);
```

not bridging the gap cleanly from a value callable into a promise interface without losing any flexibility at all?

**Comment by jaredhoberock** [comment]
We've agreed not to change P0443's design unless absolutely necessary. I think we agree that it's absolutely necessary to resolve this bug, and I would personally like to do that in a way that involves the least design churn. I worry that introducing a new concept into P0443 is too big a change, and it also seems unnecessary given that something like what we've proposed above is almost possible through P0443R6's existing design.

@LeeHowes and @dhollman : is handling `auto&&` parameters your largest concern at this point? Do you think it's possible that C++20 concepts will provide a tasteful way to disambiguate normal parameters from exception parameters? I think it may be possible, but in principle, any type can be thrown as an exception, not just `exception_ptr`. Right now, `std::future` only supports `exception_ptr`, but custom future types in principle carry any type as the error type.

**Comment by LeeHowes** [comment]
Precisely. That potential ambiguity is why it is important to have clearly separated functionality between the value and error conditions - overload resolution loses the possibility of value and error being the same. set_error/set_value or a type that has a firm distinction between the two removes that ambiguity. exception_ptr just wraps an exception, so it is not the end of the world if we special case not allowing a Future<exception_ptr>. Future itself can handle that idea, though, so it's a slight shame to disallow it just because of the way we define overloads on then_execute and friends.

My (and Facebook's as I hate to take credit for everyone's views) position on this can be summarised as:
 * We already need to add a Promise concept to P0443, or we can't call then_execute dependent on a future from an incompatible executor.  
 * We are already defining Future concepts elsewhere. 
 * Adding a ReadyFuture concept that is a SemiFuture with the addition of guaranteed non-blocking reference-returning get operations is the smallest change to P0443, given that the current text already parameterises the continuation with a Future. 
 * A ReadyFuture concept is potentially the most efficient to implement because the executor can define its own type referencing or at least utilising the source Future's state directly without any conversions, which in long future chains of compatible types could be significant.
 * ReadyFuture guarantees least ambiguity in terms of necessary SFINAE and overload resolution for the implementor of a callback, therefore it is also the simplest solution. We have the ability to use an auto parameter with a reliable interface.

**Comment by ccmysen** [comment]
a couple things:
@jaredhoberock does your approach then basically mean that the successor bulk functions necessarily all need to be able to handle whatever the first exception is upstream? this could mean that if you use bulk to simulate a set of operations on independent (or semi-independent) chunks of data, you cannot propagate an exception in a conditional way for some branches. is that correct or is there another approach?
@LeeHowes i think one real worry is that if futures as described does not make it far enough along, we basically are stuck on executors as well -- question is how critically do we want to tie the forward progress of executors to the final (vs. an intermediate) desired state for futures? My take would be that we would want the ability to have something simpler which doesn't rely so much on significant changes to how futures work in order to make forward progress unless we can get futures to a happy state super quickly...

**Comment by LeeHowes** [comment]
We already have to change what is in there to make it not depend on a TS. I would suggest that if that is a concern, given that both twoway_execute and then_execute (+bulk forms) fundamentally need to take and return a future of some sort, and in writing that up we need to define what that future is if anyone is expected to use them in any remotely generic context, we should put consider merging oneway_executor and the query components immediately into the IS, and add the future dependent concepts after we agree what a future is.

**Comment by dhollman** [comment]
> We've agreed not to change P0443's design unless absolutely necessary.

While I agree with this sentiment (and pushed for it heavily at the face to face earlier this month), I think we have to be careful about taking it as a serious mandate, and I know there are committee members who would be appalled with this way of thinking about something that hasn't even been to LWG yet, even if it has been 5+ years in the making. These are the kinds of things LEWG tasked us with finding through implementation experience. I think we found one, so we can report back a "mission accomplished" to LEWG. I know it's a subjective question degrees, but IMHO the overload/`constexpr if` solution is awkward enough to warrant a change, and I suspect that enough people will agree with that position such that making the change will be the path of least resistance.

>  i think one real worry is that if futures as described does not make it far enough along, we basically are stuck on executors as well -- question is how critically do we want to tie the forward progress of executors to the final (vs. an intermediate) desired state for futures?

But this change is completely independent of the futures paper. Users of `then_execute` need to be able to handle exceptions, and if it's too awkward we'll meet some resistance regardless of the completeness of the futures paper. This is really a tiny change; most executors can directly use their `executor_future_t` for their `ready_future_t` if we're careful in how we define it, and there's so little code dependent on the parameter type of the closure to `then_execute` that most of us on the call didn't know it off of the top of our heads.

I'm totally for finding the path of least resistance to getting executors into the standard, and I honestly think this sort of change is going to be that path in this case.

**Comment by jaredhoberock** [comment]
> @jaredhoberock does your approach then basically mean that the successor bulk functions necessarily all need to be able to handle whatever the first exception is upstream? this could mean that if you use bulk to simulate a set of operations on independent (or semi-independent) chunks of data, you cannot propagate an exception in a conditional way for some branches. is that correct or is there another approach?

Whatever approach we decide to take, I would like for handling incoming exceptions to be completely optional. By this I mean two things: 1. the user function should be able to completely ignore an incoming exception (or selectively filter exceptions it chooses to handle using the C++ type system) and 2. the executor should be able to decline to present an incoming exception to the user function. In both cases, the user function would not be called at all.

**Comment by LeeHowes** [comment]
2) 
This is easy, and intended in our original proposal on this. I'd like to see it queryable on the Executor, so we might want to consider adding a known query for this feature.

1) 
This is quite expensive in the general case because of the cost of catch + rethrow in an exception chain. It is less of a problem if we can list a set of exceptions that can be captured. In folly we wrap exception_ptr in exception_wrapper to reduce the high cost in .onError chains. For this reason we think the primary mode here should be to just pass the exception_ptr through and let a try...catch...catch sequence in the code see it.

However, I do see good justification for supporting what you want, because for bulk execution in particular it would be nice to filter once rather than in each worker. 

I see two obvious solutions to this. The first is to add something like the shape factory function I proposed earlier, which could reduce the bulk range to 0 and also do exception preprocessing. It's not quite encoded in the type, but it is in a compiler visible code block. It would open the risk of arbitrary pre-work code running in bulk executions. As Chris M pointed out though there might be some benefit to doing pre- and post-processing steps in bulk runs anyway, so I'd like us to at least consider that option.

The second is a solution that means we don't have to change the paper beyond what I suggested above. If an executor has to export its ready_future type (which makes sense, right?) and it is a template export then we can optionally extend it. So the paper just requires an export of the form:
```
template<class value_t>
using ready_future = /* implementation-defined */

```
where we define it such that "the first template argument represents the value type of the ready future". That text does not prohibit your executor from adding additional executor-specific template arguments, and the obvious one in that case is a list of valid exception types. So we end up with a callback of the form:

`[](Executor::ready_future<int, std::logic_error>&){...}`

and all but logic_error bypass. The default here should be "all" of course, unless the executor itself does not deal with exceptions at all. Now it's in the type of the function as you requested.

**Comment by LeeHowes** [comment]
exception_ptr just wraps the exceptions, though. If the Future is going to be polymorphic over its exceptions and store whatever was thrown rather than have it blow up out of the exception chain, then we need to store the exception somehow. A fully type safe get_exception (that isn't templated and doing a throw/catch internally) would be limiting the set of exceptions in the future, and that has been objectionable when used on function signatures.

So I think storage-wise we can't really avoid exception_ptr. We can extract the exception type from that, but the only well-defined way to do that is to rethrow and catch, which is expensive. In folly we use some dynamic cast and pointer storing in exception_wrapper, but I don't know if it fully generalises.

Allowing the executor to directly filter based on the type of the callee makes a lot of sense to me though. It can do the inefficient thing, or can presumably find a way to optimise it knowing the set of exceptions that can be passed in.

**Comment by jaredhoberock** [comment]
To provide context for the upcoming 2018/04/20 teleconference, I'd like to summarize what I believe is the consensus among @jaredhoberock, @LeeHowes, and @dhollman at this point in the thread. 

We agree that the issue described in the original post must be resolved for our Rapperswil submission.

We agree that `bulk_then_execute`'s user-provided parameter `f` should be the object responsible for implementing the bulk continuation function that either consumes a predecessor value `v` or optionally handles a predecessor exception `e`.

We agree that executor types should be able to opt-out of allowing `f` to handle a predecessor exception. In such a case, a predecessor exception may be forwarded to `bulk_then_execute`'s result future. 

I think this consensus can be summarized in these revised operational semantics for `bulk_then_execute`:

> If R is non-`void`, invokes `rf()` at most once on an unspecified execution agent to produce the value `r`. Invokes `sf()` at most once on an unspecified execution agent to produce the value `s`.
>
> When `fut` is ready, if it is non-exceptional, creates a group of execution agents of shape `n` which invokes `NORMAL_CASE(DECAY_COPY(std::forward<F>(f)), i, v, r, s)` if `R` is non-`void`, and otherwise invokes `NORMAL_CASE(DECAY_COPY(std::forward<F>(f)), i, fut, s)`, at most once for each value of `i` in the range `[0,n)`, with the call to `DECAY_COPY` being evaluated in the thread that called `bulk_then_execute`.
>
> Otherwise, if it is exceptional, optionally creates a group of execution agents of shape `n` which invokes `EXCEPTIONAL_CASE(DECAY_COPY(std::forward<F>(f)), i, e, r, s)` if `R` is non-`void`, and otherwise invokes `NORMAL_CASE(DECAY_COPY(std::forward<F>(f)), i, e, s)`, at most once for each value of `i` in the range `[0,n)`, with the call to `DECAY_COPY` being evaluated in the thread that called `bulk_then_execute`.

I believe remaining disagreement is over what should be substituted in place of `NORMAL_CASE` and `EXCEPTIONAL_CASE` in these operational semantics. There have been a few suggestions so far in this thread:

* `f.set_value(i, v, r, s)`
* `f.set_exception(i, e, r, s)`
* `f(i, v, r, s)`
* `f(i, e, r, s)`
* `f(exception_arg, i, e, r, s)`

**Comment by LeeHowes** [comment]
I would add that almost all of this discussion applies to non-bulk too, barring the reference semantics of the incoming data.

In addition, the final option, which as I said above we prefer, would be that both NORMAL_CASE and EXCEPTIONAL_CASE are simplified to:

> If R is non-void, invokes rf() at most once on an unspecified execution agent to produce the value r. Invokes sf() at most once on an unspecified execution agent to produce the value s.
> 
> When fut is ready, if it is non-exceptional, creates a group of execution agents of shape n which invokes f(DECAY_COPY(std::forward<F>(f)), i, rf, r, s) if R is non-void, and otherwise invokes f(DECAY_COPY(std::forward<F>(f)), i, rf, s), in both cases where rf is a ReadyFuture for which rf.has_value() returns true at most once for each value of i in the range [0,n), with the call to DECAY_COPY being evaluated in the thread that called bulk_then_execute.
> 
> Otherwise, if fut is exceptional, optionally creates a group of execution agents of shape n which invokes f(DECAY_COPY(std::forward<F>(f)), i, rf, r, s) if R is non-void and otherwise invokes f(DECAY_COPY(std::forward<F>(f)), i, rf, s), where in both cases rf is a ReadyFuture for which rf.has_exception() returns true. 
> In either the void or non-void case, the expression is invoked at most once for each value of i in the range [0,n), with the call to DECAY_COPY being evaluated in the thread that called bulk_then_execute.


In the pull request I worded this slightly differently as: 

> and which constructs rfr, a ReadyFuture of type T from the value or exception stored in fut,

**Comment by dhollman** [comment]
A few questions/nitpicks/etc.:

- I'd prefer that the `EXCEPTIONAL_CASE` gets invoked only once (not once for each `i`), at least by default. I don't really see the use case for the latter.
- Why is the `NORMAL_CASE` invoked "at most once" for each `i` rather than exactly once?
- "When `fut` is ready" should probably be worded in some way as to convey that this is the earliest time (maybe just "After `fut` is ready")

**Comment by ccmysen** [comment]
If i'm understanding correctly we have basically 2 things in our consensus that I worry about the interplay of:

1. That the executor may chose not to do anything with exceptions and therefore can call terminate. We probably want a property to advertise this fact.
2. That some user code may chose to pass a call which only processes the value, the exception, or both with some expectation about what happens when either exceptions or values are not part of the interface. But, even if an exception handling function is provided, the executor is not guaranteed to pass exceptions onto this function even if an exception happened in the predecessor.

My main concern is that if an executor doesn't actually say if it terminates doesn't give the user a whole lot of information because generic code would have to assume that some executors will and either has to guarantee no code will except and therefore doesn't care about the terminate properties, or code will have to somehow handle all exceptions.

the problem is that given that code has to handle all exceptions in some cases, but the executor chooses to not pass the execption to the successor function, the user would be forced to process the exception within the predecessor to be defensive in generic code.

Of course if you know your executor and its semantics (which is likely the more common case) this isn't an issue, but it would be an issue in library code which is trying to have well defined semantics...

I would suggest we might want to make some sort of guarantee on whether an exception will propagate, though would be interested to hear executors which would be unable to do this (or at least unable to do it efficiently).

**Comment by jaredhoberock** [comment]
I don't think we have 1. in our consensus. Either the executor must propagate an incoming exception to `bulk_then_execute`'s output future, or the executor must present the incoming exception to the user function. Here's the consensus I recorded in the 2018-04-20 meeting minutes:

> jared: consensus:
> 
> executor gets to decide between propagation & handling
> user function gets to decide whether it wants to handle the normal case
> user function gets to decide whether it wants to handle the exceptional case
> user function uses the call operator for the normal case
> user function uses the call operator w/ disambiguator arg for the exceptional case
> .then_execute() should work analogously
> 
> jared: XXX i'll write up a PR with this consensus and ask david & lee to review

One reason I think it's important for the executor to be able to choose between propagation and handling is that some executors will not be able to deal with `std::exception_ptr`. This is because the way `std::exception_ptr` must ultimately be dealt with is via `std::rethrow_exception(e_ptr)`. Certain kinds of execution agents (e.g. GPU agents, SIMD agents) will not be able to call `std::rethrow_exception`.

I think this would be less of a problem with other exception types that would not require additional `throw`s for handling. However, with either `std::future` or `std::experimental::future` as our only valid `Future` types, we are stuck with `std::exception_ptr` for the moment.

**Comment by LeeHowes** [comment]
> I don't think we have 1. in our consensus. Either the executor must propagate an incoming exception to bulk_then_execute's output future, or the executor must present the incoming exception to the user function.

True, but I think Chris has a point. As an author of generic code, I need to be able to ban executors that don't offer me the exception (unless there is huge value - for example because I need huge GPU parallelism in a specific case). Otherwise the behaviour of my error handling code is broken. We do need a compile-time query of some sort.

To your concerns:
To some extend this could be mitigated with explicitly named functions, and is one good reason for preferring explicitness at this level. Even then not entirely. The advantage would be that we would select a function by name, not by overload, and then fail to compile against it.

So the question is, can we achieve the same thing here? Can we say "I found a function that does not satisfy the F(exception_arg, T) signature for some T, therefore compile against it." Similarly "I found a function that does satisfy the F(exception_arg, T) signature for some T, therefore compile against it". Once we have one, we should force compiling against it, otherwise some relatively small mistake could skip, as you point out. This applies just as strongly to the exception variant, because I don't want to accidentally not write an exception handler. That could be a serious production bug and hard to catch at code review time.

I think that should be achievable. Can we do that by making NORMAL slightly more generic to make sure it catches? Something like:

> Let NORMAL be the expression f(v) for v of some type T and EXCEPTIONAL be the expression f(exception_arg, e) for std::exception_ptr e.

Then the rest of your text stays the same, but is more strict than the definition of NORMAL, so NORMAL should succeed but the actual compilation of the call should fail giving us the error we want.

**Comment by LeeHowes** [comment]
As an aside, what does the call f(v) look like when v is of type void?

**Comment by jaredhoberock** [comment]
> As an aside, what does the call f(v) look like when v is of type void?

It should look like `f()`. I've updated my PR to account for this case.

**Comment by dhollman** [comment]
This all looks good to me. I feel like Jared's concern can be addressed by LWG wording experts if it is still a problem by the time it makes it that far. I'm not an expert on such things, but I would be surprised if named functions changed that situation.

**Comment by jaredhoberock** [comment]
Fixed by R7.
